esp8266/net.go

```diff
@@ -0,0 +1,143 @@
+	addr := raddr.IP.String()
+	sendport := strconv.Itoa(raddr.Port)
+	listenport := strconv.Itoa(laddr.Port)
+
+	// disconnect any old socket
+	d.DisconnectSocket()
+
+	// connect new socket
+	d.ConnectUDPSocket(addr, sendport, listenport)
+
+	return &SerialConn{Adaptor: &d, laddr: laddr, raddr: raddr}, nil
+}
+
+// ListenUDP listens for UDP connections on the port listed in laddr.
+func (d Device) ListenUDP(network string, laddr *UDPAddr) (*SerialConn, error) {
+	addr := "0"
+	sendport := "0"
+	listenport := strconv.Itoa(laddr.Port)
+
+	// disconnect any old socket
+	d.DisconnectSocket()
+
+	// connect new socket
+	d.ConnectUDPSocket(addr, sendport, listenport)
+
+	return &SerialConn{Adaptor: &d, laddr: laddr}, nil
+}
+
+// SerialConn is a loosely net.Conn compatible intended to support
+// TCP/UDP over serial.
+type SerialConn struct {
+	Adaptor *Device
+	laddr   *UDPAddr
+	raddr   *UDPAddr
+}
+
+// Read reads data from the connection.
+// TODO: implement the full method functionality:
+// Read can be made to time out and return an Error with Timeout() == true
+// after a fixed time limit; see SetDeadline and SetReadDeadline.
+func (c *SerialConn) Read(b []byte) (n int, err error) {
+	// read only the data that has been received via "+IPD" socket
+	return c.Adaptor.ReadSocket(b)
+}
+
+// Write writes data to the connection.
+// TODO: implement the full method functionality for timeouts.
+// Write can be made to time out and return an Error with Timeout() == true
+// after a fixed time limit; see SetDeadline and SetWriteDeadline.
+func (c *SerialConn) Write(b []byte) (n int, err error) {
+	// specify that is a data transfer to the
+	// currently open socket, not commands to the ESP8266.
+	c.Adaptor.StartSocketSend(len(b))
+	return c.Adaptor.Write(b)
+}
+
+// Close closes the connection.
+// Currently only supports a single Read or Write operations without blocking.
+func (c *SerialConn) Close() error {
+	c.Adaptor.DisconnectSocket()
+	return nil
+}
+
+// LocalAddr returns the local network address.
+func (c *SerialConn) LocalAddr() UDPAddr {
+	return *c.laddr
+}
+
+// RemoteAddr returns the remote network address.
+func (c *SerialConn) RemoteAddr() UDPAddr {
+	return *c.laddr
+}
+
+// SetDeadline sets the read and write deadlines associated
+// with the connection. It is equivalent to calling both
+// SetReadDeadline and SetWriteDeadline.
+//
+// A deadline is an absolute time after which I/O operations
+// fail with a timeout (see type Error) instead of
+// blocking. The deadline applies to all future and pending
+// I/O, not just the immediately following call to Read or
+// Write. After a deadline has been exceeded, the connection
+// can be refreshed by setting a deadline in the future.
+//
+// An idle timeout can be implemented by repeatedly extending
+// the deadline after successful Read or Write calls.
+//
+// A zero value for t means I/O operations will not time out.
+func (c *SerialConn) SetDeadline(t time.Time) error {
+	return nil
```

Comment:
I wonder whether it would be better to either not implement these methods or always return an error? That makes clear these functions are not (yet) implemented.

Reply:
I left them like this, so that it would be easier to test if it was conforming with the real interface more easily. Also so that we could possibly implement them correctly in context in a future iteration.